tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ExprOptimizer.scala

```diff
@@ -147,19 +134,18 @@ class ExprOptimizer(nameGen: UniqueNameGenerator, tracker: TransformationTracker
    * @return a transformed \exists expression
    */
   private def transformExistsOverSets: PartialFunction[TlaEx, TlaEx] = {
-    case OperEx(TlaBoolOper.exists, NameEx(x), OperEx(TlaSetOper.filter, NameEx(y), s, e), g) =>
+    case OperEx(TlaBoolOper.exists, xe @ NameEx(x), OperEx(TlaSetOper.filter, ye @ NameEx(y), s, e), g) =>
       // \E x \in {y \in S: e}: g becomes \E y \in S: e /\ g [x replaced with y]
-      val newPred = ReplaceFixed(tracker)(NameEx(x), NameEx(y)).apply(tla.and(e, g))
-      val result = tla.exists(tla.name(y), s, newPred)
+      val newPred =
+        ReplaceFixed(tracker)(replacedEx = xe, newEx = NameEx(y)(ye.typeTag)).apply(tla.and(e, g))
+      val result = OperEx(TlaBoolOper.exists, NameEx(y)(ye.typeTag), s, newPred)(boolTag)
       transformExistsOverSets.applyOrElse(result, { _: TlaEx => result }) // apply recursively to the result
 
-    case OperEx(TlaBoolOper.exists, NameEx(boundVar), OperEx(TlaSetOper.map, mapEx, varsAndSets @ _*), pred) =>
-      // e.g., \E x \in {e: y \in S}: g becomes \E y \in S: LET tmp == e IN g [x replaced with e]
-      // LET x == e IN g in the example above
-      val letName = nameGen.newName()
-      val newPred = ReplaceFixed(tracker)(NameEx(boundVar), tla.appOp(NameEx(letName))).apply(pred)
+    case OperEx(TlaBoolOper.exists, xe @ NameEx(_), OperEx(TlaSetOper.map, mapEx, varsAndSets @ _*), pred) =>
+      // e.g., \E x \in {e: y \in S}: g becomes \E y \in S: g[x replaced with e]
+      // g[x replaced with e] in the example above
+      val newPred = ReplaceFixed(tracker)(replacedEx = xe, newEx = DeepCopy(tracker).deepCopyEx(mapEx)).apply(pred)
```

Comment:
@Kukovec does it look better now? I think `deepCopy` should be the default behavior on every replacement. Are there any scenarios where this should not be the case?

Reply:
Ah, missed this by 42s apparently. See my other comment.